### ISSUE TITLE
sql: the number of buckets in a hash-sharded index should be configured in a storage parameter

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#75243]
Currently, the number of buckets for a hash-sharded index is configure with the syntax:

```sql
CREATE INDEX ON t (a) USING HASH WITH BUCKET_COUNT = 12
```

A better alternative syntax would be to specify the bucket count in [storage parameters](https://www.postgresql.org/docs/14/sql-createindex.html#SQL-CREATEINDEX-STORAGE-PARAMETERS). For example:

```sql
CREATE INDEX ON t (a) USING HASH WITH (bucket_count = 12)
```

This is more idiomatic syntactically, and it feels natural because `bucket_count` is similar to the types of storage parameters supported by Postgres.

If the `bucket_count` storage parameter is supplied without `USING HASH`, an error would be thrown:

```sql
-- ERROR:  22023: unrecognized parameter "bucket_count"
-- HINT: bucket_count is only allowed for hash-sharded indexes created with USING HASH
CREATE INDEX ON t (a) WITH (bucket_count = 12)
```

Once we've fully removed the current syntax, we'll no longer need a lookahead hack in the lexer to support the soon-to-be-added syntax `CREATE INDEX ON t (a) USING HASH`, because the option `WITH BUCKET_COUNT ...` will no longer be ambiguous with the optional `WITH` in storage parameters.

Ultimately, all the variants of hash-sharded indexes would be:

```sql
CREATE INDEX ON t (a) USING HASH
CREATE INDEX ON t (a) USING HASH WITH (bucket_count = 12)
```

Epic: CRDB-7363

**Comment by mgartner** [comment]
Another consideration here is the `opt_storing` and `opt_paritition_by_index` clauses that sit between `opt_hash_sharded` and `opt_with_storage_parameter_list`. https://github.com/cockroachdb/cockroach/blob/eb26eb323b14c2e00d2bf02bbd02cc7f84ef2a93/pkg/sql/parser/sql.y#L7039

If we leave as-is, the syntax required to make a hash-sharded index with a custom bucket count and storing columns would be:

```sql
CREATE INDEX ON t (a) USING HASH STORING (b) WITH (bucket_count = 12)
```

In my opinion, `STORING (b)` breaking up `USING HASH` and `WITH ...` is aesthetically fine and doesn't add confusion, but other may feel differently.

**Comment by vy-ton** [comment]
For backward compatibility, we should keep the current syntax too and not force a rewrite in this scenario. FYI @devadvocado

**Comment by chengxiong-ruan** [comment]
This fixed with these prs : #75971 #76068 #76112
